src/mango/depository.ts

```diff
@@ -290,4 +301,83 @@ export class MangoDepository {
+    const fees = realAmount * takerFee;
+    const slippage = (perfectAmount - realAmount) - fees;
+    const finalYield = realAmount - fees;
+    return { yield: finalYield, fees, slippage };
+  }
+
+  // Estimated amount of Collateral (SOL/BTC/ETH) that will be given for Redeeming (Redeem or RebalancingLite when PnlPolarity is negative)
+  // `collateralQuantity` : In UXD UI units
+  //  Output -> Collateral UI Units, estimated
+  public async getRedeemingEstimates(
+    redeemableQuantity: number,
+    perpPrice: number,
+    priceImpact: number,
+    mango: Mango
+  ): Promise<SwapEstimate> {
+    const takerFee = this.getCollateralPerpTakerFees(mango);
+    const collateralQuantity = redeemableQuantity / perpPrice;
+    const perfectAmount = collateralQuantity * perpPrice;
+    const realAmount = collateralQuantity * priceImpact;
+    const fees = realAmount * takerFee;
+    const slippage = (perfectAmount - realAmount) - fees;
+    const finalYield = realAmount - fees;
+    return { yield: finalYield, fees, slippage };
+  }
+
+  // Price impact for a minting operation (Mint or RebalancingLite when PnlPolarity is positive)
+  public async getMintingPriceImpact(
+    collateralQuantity: number,  // UI units
+    mango: Mango
+  ): Promise<number | undefined> {
+    // Minting is placing a ask IoC limit order, will check the impact on the bid side
+    return this.getMangoPerpOrderBookPriceImpact(mango, collateralQuantity, "ask");
+  }
+
+  // Price impact for a redeeming operation (Redeeming or RebalancingLite when PnlPolarity is negative)
+  public async getRedeemingPriceImpact(
+    redeemableQuantity: number, // UI units
+    perpPrice: number,
+    mango: Mango
+  ): Promise<number | undefined> {
+    const takerFee = this.getCollateralPerpTakerFees(mango);
+    // Remove the max possible fees, as we place an order on that amount to  make sure we can pay for the fees on the program side.
+    const redeemableQuantityMinusFees = redeemableQuantity * (1 - takerFee);
+    const collateralQuantity = redeemableQuantityMinusFees / perpPrice;
+    // Redeeming is placing a bid IoC limit order, will check the impact on the ask side
+    return this.getMangoPerpOrderBookPriceImpact(mango, collateralQuantity, "bid");
+  }
+
+  // Return the price impact for an order on a 
+  public async getMangoPerpOrderBookPriceImpact(
+    mango: Mango,
+    quantity: number, // UI AMount
+    side: "bid" | "ask"
+  ): Promise<number | undefined> {
+    const pmc = this.getPerpMarketConfig(mango); // perpMarketConfig
+    const pm: PerpMarket = await mango.client.getPerpMarket(
+      pmc.publicKey,
+      pmc.baseDecimals,
+      pmc.quoteDecimals
+    ); // perpMarket
+    const nativeQuantity = (new BN(quantity)).mul(new BN(this.collateralMintDecimals).pow(new BN(10)));
+
+    const loadSideBook = side === 'bid' ? pm.loadBids : pm.loadAsks;
+    const sideBook = await loadSideBook(mango.client.connection, false);
```

Comment:
The problem here is that you need to retrieve the `PerpMarket` through `getPerpMarket` before being able to load the `SideBook` through `loadSidebook`.

I assume each of these query the chain through RPC calls.

Ideally, we would be able to retrieve the necessary info with a single RPC call, but here we are limited by the API offered by `mango-client`.

We could open a PR at `mango-client` to try & change that, but it's not sure it's feasible anyway.

What we can do on the other hand, is to retrieve once & keep a `PerpMarket` instance around, that we can re-use  as much as possible, so we only call `loadSideBook` here.

I don't think we should do this right now though, it's better to ship the feature first then optimize later, because it'll lead to potentially a lot of refactoring/changes.

Although it's important indeed we keep performance aspects in mind.

Also took the liberty to slightly rewrite this function to re-use existing helpers:

```ts
  public async getMangoPerpOrderBookPriceImpact(
    mango: Mango,
    quantity: number, // UI AMount
    side: 'bid' | 'ask'
  ): Promise<number | undefined> {
    const nativeQuantity = uiToNative(quantity, this.collateralMintDecimals);

    const { loadBids, loadAsks } = await this.getPerpMarket(mango);
    const loadSideBook = side === 'bid' ? loadBids : loadAsks;
    const sideBook = await loadSideBook(mango.client.connection, false);

    return sideBook.getImpactPriceUi(nativeQuantity);
  }
```

Reply:
Understood, thanks for the feedback will include this change